### ISSUE TITLE
Cosmetic - strange units and x-axis not long enough on SPSA graph for running tune

### ISSUE THREAD

**Issue by xoto10** [glinscott/fishtest#1011]
For example, I am looking at https://tests.stockfishchess.org/tests/view/6130936b45091e810014af76 and it shows 9024 games played:
![image](https://user-images.githubusercontent.com/23479932/131846313-02294aa6-f10c-4ab3-90eb-e9286f316f6a.png)
but the x-axis on the graph only goes to 10 (what units?) and appears to miss out some data at the end:
![image](https://user-images.githubusercontent.com/23479932/131846769-dae3e362-6b56-4d87-9b6d-5b449d1f12fb.png)
Note that at this point the current value for nn_scale is now 6926
![image](https://user-images.githubusercontent.com/23479932/131846914-a8daff5d-fc76-4e35-a8e6-c28275846cd6.png)

I have found https://developers.google.com/chart/interactive/docs/gallery/linechart and https://github.com/glinscott/fishtest/blob/6d9dda90463e6c6f53a629a9f344597fa25d2bde/server/fishtest/static/js/spsa.js but am still figuring out how this all works ...

**Comment by xoto10** [comment]
Maybe
`'explorer': {}`
could be added to the chart_options near https://github.com/glinscott/fishtest/blob/6d9dda90463e6c6f53a629a9f344597fa25d2bde/server/fishtest/static/js/spsa.js#L28

Hopefully there isn't much downside from trying that?

**Comment by ppigazzini** [comment]
@xoto10 PROD updated by hand, test the new behavior

**Comment by xoto10** [comment]
Thanks @ppigazzini ! 
Ok, that doesn't solve anything, but it does allow the graph to be zoomed, which shows there is no "missing" data off the right-hand side of the graph, which is what I was wondering about. I don't think it does any harm so we could leave this in for the moment.

The problem seems to be that e.g. the tune i am looking at has processed ~8700 iterations, but the graph only has 20 data points, so there is some sampling going on somewhere, and presumably the graph is only displaying data every N iterations (N maybe ~400 in this case). Considering that a tune might have 100,000 iterations, it makes sense to sample the data for the graph, but I guess I have 3 issues with this:
- a finer grain would be nice, say ~100 data points instead of ~20
- it would be nice to see the latest data point on the rhs of the graph, so if there have been 8750 iterations and we're displaying a data point for every 400, I would prefer to see iteration 8750 for the last point, not iteration 8400
- it's not clear what the translation is from iteration number to label on the x-axis

Perhaps this is all not very important, it would just be nice to understand what is happening a bit better, e.g. how the numbering on the x-axis is computed, it is not clear what value displayed on the x-axis corresponds to what iteration number. I will have a closer look to see if I can see where the sampling is done, or if the x-axis labelling can be modified to make it clearer.

**Comment by ppigazzini** [comment]
IIRC there is also the open problem about showing a single negative parameter: the zoom factor is suboptimal because the graph contains also the x-axis.

**Comment by xoto10** [comment]
Ah ok https://github.com/glinscott/fishtest/blob/8dc945edec22699f1de6c7ba583892b00989e30b/server/fishtest/rundb.py#L1053

My example has 17 params and I mentioned ~400 above, the code does 17*25 giving 425, matching my 400 estimate. Now I know the code to look at ... :-)

**Comment by xoto10** [comment]
@SFisGOD any thoughts on this, or do you not worry about this?

Ok, so the code takes the number of parameters being tuned and multiplies by 25, generally using this number for the frequency of updates (in iterations).

Regarding the 3 points I mentioned earlier, my current thoughts are:

- The finer grain is not really an issue, it just indicates a test is early in its run.

- Seeing the latest data point is not an issue during a test, we just have to wait a while for the next update.

- Perhaps this code is a bit hacky, but I have a suggestion for making the translation from x-axis label to iteration number a bit more intuitive, and also to ensure that it divides exactly into the total number of iterations (assuming most tests will be a multiple of 10000 iterations). A more complete solution would be to store an index number or iteration number directly in the spsa summary, but this would be more work, and would also make these data files larger. My suggestion is for line 1059 of rundb.py :
```
          freq = L * 25
+         # adjust freq to multiples of 100 that divide exactly into 10000; 100/200/400/500/1000/2000/2500/5000
+         while freq < 10000 and (10000%freq != 0 or freq%100 != 0):
+             freq = freq - 25
          if freq < 100:
```
This ensures that the translation from iteration number to x-axis label uses "round" numbers which should be more intuitive to translate into iteration numbers in the test (e.g. multiply by 200 instead of 375, 400 instead of 425, etc). Assuming the test is a multiple of 10,000 iterations it also ensures that the last data point will be displayed on the graph once the test has finished. This is not currently the case as e.g. 425 does not divide exactly into 30,000.

**Comment by SFisGOD** [comment]
@xoto10 I used to look at the these graphs to try to identify parameters that were only doing random walks but it was time consuming and I had inconsistent results. So I stopped looking at these graphs and focused on other factors that have a bigger impact.

**Comment by ppigazzini** [comment]
@xoto10 PROD updated by hand.

**Comment by xoto10** [comment]
@ppigazzini 
Haha, wasn't expecting that to be implemented so quickly, but it seems ok. Cool :-)

I've updated the issue about graphing a single negative parameter.

**Comment by xoto10** [comment]
@ppigazzini 
I notice the change to rundb.py has been reverted, e.g. my current tune at https://tests.stockfishchess.org/tests/view/613642cc0cd98ab40c0c9d8f is now using a frequency of 375 iterations to add each new point to the graph. Do you want me to create a PR for this change, or do you prefer not to implement it? (I like it, I prefer to have a multiple of 200 to a multiple of 375, but it's not a critical issue, so I'm happy either way.)

**Comment by ppigazzini** [comment]
@xoto10 the changes by hand are gone with the latest server update. Please open a PR so I can test on DEV.

**Comment by ppigazzini** [comment]
Fixed by #1018
